### ISSUE TITLE
cache_limit bottlenecks the performance

### ISSUE THREAD

**Issue by mayankjobanputra** [mosaicml/streaming#339]
## Environment

- OS: [RHEL]
- Hardware (GPU, or instance type): 8x[A100 40gb]
- ENV: SLURM single compute node

## To reproduce

1. Choose a reasonably large dataset (i.e., OSCAR)
2. Create a dataloader using StreamingDataset
3. set appropriate cache_limit (<< the size of the dataset)

## Expected behavior
StreamingDataset local cache should be managed in a way that it is always available (for new data) and doesn't affect the data loading performance when downloading from the remote. Currently, once the cache is full, the data loading becomes extremely slow and becomes a bottleneck for the training.

## Additional context

![image](https://github.com/mosaicml/streaming/assets/10355927/5590137c-e82f-42f0-89a9-2d42155c70fe)

**Comment by karan6181** [comment]
Hi @mayankjobanputra , Few questions I have to understand your use-case better in case I can provide any suggestion to improve the throughput. 
- What is the dataset size?
- What is the value of `cache_limit` are you setting?
- Is your dataset compressed ?
- What is the `shuffle_algo` you are using?
- What is the `predownload` value ?

**Comment by mayankjobanputra** [comment]
Hey @karan6181,
Thank you for responding :) Here are the answers.

> * What is the dataset size?

The dataset size is 2B words, 

> * What is the value of `cache_limit` are you setting?

my `cache_limit` is 450gb,

> * Is your dataset compressed ?

yes with zstd level 3 (default)

> * What is the `shuffle_algo` you are using?

py1b (default from llm-foundry)

> * What is the `predownload` value ?

100_000 (default from llm-foundry)

**Comment by karan6181** [comment]
Thanks for sharing the details. Few suggestions:
- Use `py1s` as a shuffling algorithm (Which is default in [StreamingDataset](https://github.com/mosaicml/streaming/blob/main/streaming/base/dataset.py#L252)). You can find more information about the different shuffling algorithms [here](https://docs.mosaicml.com/projects/streaming/en/stable/fundamentals/shuffling.html). TLDR is that py1b downloads lots of shard files upfront based on how samples are shuffled in py1b. Although, there are ways to control it, but for now, let use `py1s` and see if it improves the throughput or not!
- Use the streaming `predownload` default value or set it to `4 * per_gpu_batch_size`.

**Comment by mayankjobanputra** [comment]
Thanks a lot @karan6181  :) I will give this a go for the next run :)

I am closing the issue right now because it doesn't make sense to keep it open for about 15 days. I will re-open if I need more help :)

I will post my findings anyways :)

**Comment by mayankjobanputra** [comment]
@karan6181 I was able to increase my cache_limit to 900gb and now with `py1b` it worked fine for the entire duration. Everything else was the same as before.

I will try to stick to `py1b` for now, but I will definitely try `py1s` in the near future. Right now it might yield a little different results which is not ideal for my experiments. Thanks again for your help :)

P.S. I love Streaming docs :tada:  You guys have put really enough details to infer the right choices. Kudos to the team!

**Comment by mayankjobanputra** [comment]
I think I made a small blunder. It seems like although I have been passing `shuffling_algo` all this while, the `shuffle` was set to `False` which means no shuffling was going on.

@karan6181 Do you suggest turning on the shuffling? Will it have any benefits? All I am looking for is performance and stability.

**Comment by karan6181** [comment]
@mayankjobanputra Do you care convergence ? If yes, the I would recommend to do `shuffle=True` !

**Comment by mayankjobanputra** [comment]
@karan6181 You are right, I do see not-so-great convergence in my training loss (small spikes everywhere), but I am not sure how that would fix the original performance bottleneck that probably occurs primarily because of the cache_limit. 

Would you mind elaborating? I am quite new to large-scale pertaining stuff, so please forgive my naive questions.

**Comment by karan6181** [comment]
How is the convergence performance when `shuffle_algo=py1s` and `shuffle=True` vs `shuffle_algo=py1b` and `shuffle=True` with `cache_limit= 900gb` ?

**Comment by mayankjobanputra** [comment]
So far I haven't had any model using `shuffle=True` so I really can't comment on that. So far, I have been using defaults which have `shuffle=False` and the looks something like this.

![image](https://github.com/mosaicml/streaming/assets/10355927/0ae477a3-b737-4d66-a3e6-a5dda2376c5f)
![image](https://github.com/mosaicml/streaming/assets/10355927/380570ab-936a-4d24-8458-1be6c6acf6a9)

**Comment by karan6181** [comment]
Looks like the convergence is kind of the same but this is with `shuffle=False` though. With `shuffle=True`, I would recommend `shuffle_algo=py1b` with default predownload (you can experiment with lowering the predownload to `2 * per_gpu_batch_size` or even `1 * per_gpu_batch_size`), lower the `num_canonical_nodes` to `16 * physical nodes` or even `4 * physical nodes`), lower the `shuffle_block_size` by a factor of two. The idea behind lower certain parameter is that lower `num_canonical_nodes` you have with smaller `shuffle_block_size` then more shards are in common between your workers on different devices which leads to less downloads. However, data diversity might get impacted. Hence, you need to tweak certain parameters to find the best spot based on the dataset you have.

**Comment by mayankjobanputra** [comment]
Thanks a lot Karan :) I will try out your advice soon. For now I am closing the issue again :see_no_evil:
